src/OpenTelemetry/Trace/Span.cs

```diff
@@ -69,7 +69,7 @@ public sealed class Span : ISpan
                 parentSpanContext,
                 name,
                 null,
-                links, // we'll enumerate again, but double enumeration over small collection is cheaper than allocation
+                linksProvider,
```

Comment:
@lmolkova we'll pass the `Func<>` here, which will be invoked buy the sampler.
If it returns true, we'll need to invoke it again a bit below.

I guess the trade off here is: 2 allocations 100% of times (`LinksProvider` and `Func<>`) that go to waste when things are not sampled.
Or having just 1 allocation in all cases, the `Fun<>`, and when Spans **are** sampled only, it'll be invoked twice so if realizing the `Func<>` means deserialization, it'd be a lot more expensive. But that that run only in sampled Spans.

Reply:
Sampler should get a list of links, not a func,, so span ctor will invoke func and enumerate links only once

Reply:
Right! Span is in the SDK so that makes sense.